core/src/main/scala/org/apache/spark/mllib/evaluation/RichBinaryClassificationMetrics.scala

```diff
@@ -0,0 +1,285 @@
+/*
+ * Modified on 7/20/2020 to expose confusion matrices by threshold.
+ */
+
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.mllib.evaluation
+
```

Comment:
why is this in the spark namespace? 

I am really unhappy with just copy pasting the whole class. It is an incredibly bad idea as it gives up a huge amount of code that we are responsible for maintaining feature parity on. 

It would be better to only copy the part that we need even if it means the calculations for that will be done twice when we use the original class

Reply:
It's in the Spark namespace because it uses other classes that are package private to `evaluation`. I couldn't copy out just the piece of code that does the confusion matrix calculation into our namespace since that piece relies on Spark's `BinaryLabelCounter` and `BinaryConfusionMatrix` classes. 

You'd really rather have me copy _just_ the `confusions` (which still needs to be in `org.apache.spark.mllib.evaluation`), and do the entire calculation twice? It seems wasteful to me to just get the benefit of avoiding a few extra pieces of the Spark class, but ok.

Reply:
There is a tradeoff between computational efficiency and code that will need to be maintained. I would prefer to have less copy pasted code over. I am fine with doing the calculations twice to get that

Reply:
There's an additional problem with partially copying this class. If we still get the rest of the binary classification metrics from a call to Spark's class, then we now rely on two implementations of the threshold calculations and these could diverge.

If we copy over everything that contributes to any of the threshold metrics we use so that all our metrics are consistent, then that's basically copying the entire class! Even the AuROC and AuPR calculations we use depend on the confusion matrices by threshold. 

I share the desire to minimize any copy/pasting from other libraries, but I don't see a way to partially copy this class without introducing problems like this.

Reply:
I changed it to extend Spark's class and just reimplement explicit confusion matrix piece, but I prefer copying the whole class. This approach still has the problems I mentioned in my comment above. I'll go with it if you and @nicodv think this is better, though.

Reply:
Thanks for clearly laying out the pros and cons, @Jauntbox . I do think this better, all things considered.

Your tests that check the reconstruction of the Spark-calculated thresholds should cover the divergence scenario, right?

Reply:
The tests don't directly check the two threshold arrays against each other since I think it's too confusing to expose both of them. They do check for the situations we want to avoid the most (where the thresholds are at "different" points in the scores in the sense of model performance metrics). You can still have cases where the thresholds would shift up/down slightly and the precision/recall/fpr tests could still pass.